File: src/xreal_imu.swift
// xreal_imu.swift — native macOS reader for the XREAL Air 2 Ultra 1 kHz IMU.
// Counterpart of python/xreal_imu.py (which also runs on macOS via hidapi);
// packet formats are documented in docs/PROTOCOL.md.
//
// Build: swiftc -O src/xreal_imu.swift -o xreal_imu -framework IOKit -framework CoreFoundation
//        (or just `make`)
// Usage:
//   ./xreal_imu                     # live console output (1 line/s)
//   ./xreal_imu --csv out.csv       # also log every sample
//   ./xreal_imu --config out.json   # dump the factory calibration JSON, exit
//   ./xreal_imu --info              # serial + firmware versions, exit
//   ./xreal_imu --seconds N         # stop after N seconds
//
// The tool broadcasts commands to all vendor HID interfaces (usage page 0x41)
// and lets the replies identify the right one — the MCU channel answers 0xFD
// packets, the IMU channel answers 0xAA packets and streams 01-02 reports.

import Foundation
import IOKit
import IOKit.hid

let VID = 0x3318, PID = 0x0426

// ---- CRC-32 (zlib) -----------------------------------------------------------
let crcTable: [UInt32] = (0..<256).map { i -> UInt32 in
    var c = UInt32(i)
    for _ in 0..<8 { c = (c & 1) != 0 ? 0xEDB88320 ^ (c >> 1) : c >> 1 }
    return c
}
func crc32(_ buf: ArraySlice<UInt8>) -> UInt32 {
    var r: UInt32 = 0xffffffff
    for b in buf { r = (r >> 8) ^ crcTable[Int((r ^ UInt32(b)) & 0xff)] }
    return r ^ 0xffffffff
}

// ---- packet builders ------------------------------------------------------------
func put16(_ p: inout [UInt8], _ v: UInt16, _ o: Int) { p[o] = UInt8(v & 0xff); p[o + 1] = UInt8(v >> 8) }
func put32(_ p: inout [UInt8], _ v: UInt32, _ o: Int) { for k in 0..<4 { p[o + k] = UInt8((v >> (8 * k)) & 0xff) } }

/// IMU-channel command: aa | crc32 | len u16 (= data+3) | cmd u8 | data
func imuPacket(_ cmd: UInt8, _ data: [UInt8] = []) -> [UInt8] {
    var p = [UInt8](repeating: 0, count: 64)
    p[0] = 0xaa
    put16(&p, UInt16(data.count + 3), 5)
    p[7] = cmd
    for (i, b) in data.enumerated() { p[8 + i] = b }
    put32(&p, crc32(p[5..<(5 + data.count + 3)]), 1)
    return p
}

/// MCU-channel command: fd | crc32 | len u16 (= data+17) | req u32 | ts u32 | cmd u16 | rsvd[5] | data
func mcuPacket(_ cmd: UInt16, _ data: [UInt8] = []) -> [UInt8] {
    var p = [UInt8](repeating: 0, count: 64)
    p[0] = 0xfd
    let length = UInt16(data.count + 17)
    put16(&p, length, 5)
    put32(&p, 0x1337, 7)
    put16(&p, cmd, 15)
    for (i, b) in data.enumerated() { p[22 + i] = b }
    put32(&p, crc32(p[5..<(5 + Int(length))]), 1)
    return p
}

// ---- little-endian readers --------------------------------------------------------
func u16(_ b: UnsafePointer<UInt8>, _ o: Int) -> Int { Int(b[o]) | Int(b[o + 1]) << 8 }
func u32(_ b: UnsafePointer<UInt8>, _ o: Int) -> UInt32 {
    UInt32(b[o]) | UInt32(b[o + 1]) << 8 | UInt32(b[o + 2]) << 16 | UInt32(b[o + 3]) << 24
}
func u64(_ b: UnsafePointer<UInt8>, _ o: Int) -> UInt64 {
    (0..<8).reduce(UInt64(0)) { $0 | UInt64(b[o + $1]) << (8 * UInt64($1)) }
}
func i24(_ b: UnsafePointer<UInt8>, _ o: Int) -> Int32 {
    let v = Int32(b[o]) | Int32(b[o + 1]) << 8 | Int32(b[o + 2]) << 16
    return (v & 0x80_0000) != 0 ? v - 0x100_0000 : v
}

// ---- args --------------------------------------------------------------------------
let args = CommandLine.arguments
func argValue(_ flag: String) -> String? {
    guard let i = args.firstIndex(of: flag), args.count > i + 1 else { return nil }
    return args[i + 1]
}
let csvPath = argValue("--csv")
let configPath = argValue("--config")
let wantInfo = args.contains("--info")
let seconds = argValue("--seconds").flatMap { Double($0) }

// ---- device plumbing ------------------------------------------------------------------
var devices: [IOHIDDevice] = []
var csvHandle: FileHandle?

// latest command replies, keyed by channel head byte
var aaReply: (cmd: UInt8, data: [UInt8])?
var fdReply: (cmd: UInt16, data: [UInt8])?

// streaming stats
var sampleCount = 0, lastPrint = Date()
var totalSamples = 0

func handleImuReport(_ r: UnsafePointer<UInt8>, _ len: Int) {
    guard len >= 58, r[0] == 0x01, r[1] == 0x02 else { return }
    let ts = u64(r, 4)
    let gm = Float(u16(r, 12)), gd = Float(u32(r, 14))
    let am = Float(u16(r, 27)), ad = Float(u32(r, 29))
    let g = (0..<3).map { Float(i24(r, 18 + 3 * $0)) * gm / gd }   // deg/s
    let a = (0..<3).map { Float(i24(r, 33 + 3 * $0)) * am / ad }   // g
    sampleCount += 1
    totalSamples += 1
    if let h = csvHandle {
        let line = "\(ts),\(u32(r, 54)),\(u16(r, 2)),"
            + String(format: "%.4f,%.4f,%.4f,%.5f,%.5f,%.5f\n",
                     g[0], g[1], g[2], a[0], a[1], a[2])
        h.write(Data(line.utf8))
    }
    let dt = Date().timeIntervalSince(lastPrint)
    if dt >= 1 {
        let mag = (a[0] * a[0] + a[1] * a[1] + a[2] * a[2]).squareRoot()
        print(String(format: "%6.0f Hz  gyro=(%+8.3f,%+8.3f,%+8.3f) deg/s  " +
                     "accel=(%+6.3f,%+6.3f,%+6.3f) g  |a|=%5.3f  t_raw=%d",
                     Double(sampleCount) / dt, g[0], g[1], g[2], a[0], a[1], a[2],
                     mag, u16(r, 2)))
        sampleCount = 0
        lastPrint = Date()
    }
}

let inputCB: IOHIDReportCallback = { _, _, _, _, _, report, length in
    switch report[0] {
    case 0x01: handleImuReport(report, length)
    case 0xaa:
        guard length >= 8 else { return }
        let l = u16(report, 5)
        aaReply = (report[7], (8..<min(length, 8 + max(0, l - 3))).map { report[$0] })
    case 0xfd:
        guard length >= 22 else { return }
        let l = u16(report, 5)
        fdReply = (UInt16(u16(report, 15)),
                   (22..<min(length, 22 + max(0, l - 17))).map { report[$0] })
    default: break
    }
}

var reportBuffers: [UnsafeMutablePointer<UInt8>] = []
let mgr = IOHIDManagerCreate(kCFAllocatorDefault, IOOptionBits(kIOHIDOptionsTypeNone))
IOHIDManagerSetDeviceMatching(mgr, [kIOHIDVendorIDKey: VID, kIOHIDProductIDKey: PID] as CFDictionary)
IOHIDManagerRegisterDeviceMatchingCallback(mgr, { _, _, _, device in
    let up = (IOHIDDeviceGetProperty(device, kIOHIDPrimaryUsagePageKey as CFString) as? Int) ?? 0
    if up != 0x41 { return }                          // vendor interfaces only
    devices.append(device)
    let buf = UnsafeMutablePointer<UInt8>.allocate(capacity: 1024)
    reportBuffers.append(buf)
    IOHIDDeviceOpen(device, IOOptionBits(kIOHIDOptionsTypeNone))
    IOHIDDeviceRegisterInputReportCallback(device, buf, 1024, inputCB, nil)
}, nil)
IOHIDManagerScheduleWithRunLoop(mgr, CFRunLoopGetCurrent(), CFRunLoopMode.defaultMode.rawValue)
IOHIDManagerOpen(mgr, IOOptionBits(kIOHIDOptionsTypeNone))
CFRunLoopRunInMode(.defaultMode, 0.5, false)
guard !devices.isEmpty else {
    FileHandle.standardError.write(Data("XREAL vendor HID interfaces not found - glasses plugged in?\n".utf8))
    exit(1)
}

func broadcast(_ packet: [UInt8]) {
    for dev in devices {
        _ = packet.withUnsafeBufferPointer {
            IOHIDDeviceSetReport(dev, kIOHIDReportTypeOutput, 0, $0.baseAddress!, packet.count)
        }
    }
}

/// Send an IMU-channel command and wait for its reply.
func imuCommand(_ cmd: UInt8, _ data: [UInt8] = [], timeout: Double = 2) -> [UInt8]? {
    aaReply = nil
    broadcast(imuPacket(cmd, data))
    let deadline = Date().addingTimeInterval(timeout)
    while Date() < deadline {
        CFRunLoopRunInMode(.defaultMode, 0.02, true)
        if let r = aaReply, r.cmd == cmd { return r.data }
    }
    return nil
}

func mcuCommand(_ cmd: UInt16, _ data: [UInt8] = [], timeout: Double = 2) -> [UInt8]? {
    fdReply = nil
    broadcast(mcuPacket(cmd, data))
    let deadline = Date().addingTimeInterval(timeout)
    while Date() < deadline {
        CFRunLoopRunInMode(.defaultMode, 0.02, true)
        if let r = fdReply, r.cmd == cmd { return r.data }
    }
    return nil
}

// ---- modes -------------------------------------------------------------------------------
if wantInfo {
    // reply payload = 1 status byte + ascii text (ids per the Air-family MCU
    // protocol, wheaney/nrealAirLinuxDriver device_mcu.h)
    for (label, cmd) in [("serial       ", UInt16(0x15)), ("MCU app FW   ", 0x26),
                         ("DP7911 FW    ", 0x16), ("DSP app FW   ", 0x21)] {
        if let d = mcuCommand(cmd), d.count > 1 {
            print("\(label): \(String(bytes: d.dropFirst(), encoding: .ascii) ?? "?")")
        } else {
            print("\(label): no reply")
        }
    }
    exit(0)
}

if let path = configPath {
    _ = imuCommand(0x19, [0x00])                       // stream off while reading
    guard let lenReply = imuCommand(0x14), lenReply.count >= 4 else {
        print("config length request failed"); exit(1)
    }
    let total = Int(u32(lenReply, 0))
    var cfg = [UInt8]()
    while cfg.count < total {
        guard let part = imuCommand(0x15), !part.isEmpty else {
            print("config read stalled at \(cfg.count)/\(total)"); exit(1)
        }
        cfg.append(contentsOf: part)
    }
    try? Data(cfg.prefix(total)).write(to: URL(fileURLWithPath: path))
    print("factory calibration (\(total) bytes) -> \(path)")
    exit(0)
}

if let path = csvPath {
    FileManager.default.createFile(atPath: path, contents: nil)
    csvHandle = FileHandle(forWritingAtPath: path)
    csvHandle?.write(Data("ts_ns,counter,temp_raw,gx_dps,gy_dps,gz_dps,ax_g,ay_g,az_g\n".utf8))
}

guard imuCommand(0x19, [0x01]) != nil else {
    print("no ack for IMU stream enable"); exit(1)
}
print("IMU stream on (1 kHz). Ctrl+C to stop.")
let deadline = seconds.map { Date().addingTimeInterval($0) }
while deadline.map({ Date() < $0 }) ?? true {
    CFRunLoopRunInMode(.defaultMode, 0.25, true)
}
csvHandle?.closeFile()
print("captured \(totalSamples) samples")

File: python/xreal_imu.py
#!/usr/bin/env python3
"""1 kHz IMU reader for the XREAL Air 2 Ultra (Windows / Linux / macOS).

The glasses stream their IMU over vendor HID interface 2. Commands on that
interface use an 0xAA-framed packet (a shorter sibling of the 0xFD MCU
protocol, see docs/PROTOCOL.md):

    aa | crc32 u32 | length u16 (= len(data)+3) | cmd u8 | data...
    crc32 = zlib CRC-32 over bytes [5 : 5+length]

  cmd 0x19 data 01/00   IMU stream on / off
  cmd 0x14              -> u32 length of the calibration JSON
  cmd 0x15              -> next chunk of the calibration JSON

Once enabled, 512-byte input reports arrive at 1000 Hz:

    [0:2]   signature 01 02
    [2:4]   u16 temperature, raw
    [4:12]  u64 timestamp [ns]
    [12:14] u16 gyro multiplier      } raw * mul / div
    [14:18] u32 gyro divisor         }   = angular rate [deg/s]
    [18:27] 3 x i24 gyro x,y,z
    [27:29] u16 accel multiplier     } raw * mul / div
    [29:33] u32 accel divisor        }   = acceleration [g]
    [33:42] 3 x i24 accel x,y,z
    [42:48] u16 mul + u32 div for the magnetometer (values read 0 here)
    [48:54] 3 x i16 mag x,y,z
    [54:58] u32 secondary counter, ~976.56 us/tick (the sensor's 1024 Hz ODR)
    rest    zero padding

Values are reported in the raw sensor frame; the factory calibration
(biases, camera<->IMU extrinsics, fisheye624 intrinsics of both tracking
cameras) is stored on the device as JSON — dump it with --config.

Usage:
  python xreal_imu.py                    # live console output
  python xreal_imu.py --csv imu.csv      # also log every sample
  python xreal_imu.py --fb               # also publish shm ring "xreal_imu_fb"
  python xreal_imu.py --config calib.json  # dump factory calibration, exit
  python xreal_imu.py --seconds 10       # stop after N seconds

Needs `pip install hidapi`. On Linux run as root or add a udev rule for
VID 3318 (see README).
"""

import argparse
import struct
import sys
import time
import zlib

import hid

VID, PID, IMU_INTERFACE, MCU_INTERFACE = 0x3318, 0x0426, 2, 0
SIG = b"\x01\x02"

# shared-memory ring (--fb): 64B header + capacity slots of 48 bytes
#   header: u32 magic 'XRIM', u32 version=1, u32 capacity, u32 slot_size,
#           u64 write_count (slots [0, write_count) are valid,
#           newest slot = (write_count-1) % capacity), rest reserved
#   slot:   u64 ts_ns, u32 counter, u16 temp_raw, u16 pad,
#           f32 gyro x,y,z [deg/s], f32 accel x,y,z [g], f32 pad
IMU_FB_NAME = "xreal_imu_fb"
IMU_FB_MAGIC = 0x4D495258
IMU_FB_CAPACITY = 4096
IMU_FB_SLOT = 48
IMU_FB_HDR = 64


class ImuSample:
    __slots__ = ("ts_ns", "counter", "temp_raw", "gyro_dps", "accel_g")

    def __init__(self, report):
        b = report
        self.temp_raw, = struct.unpack_from("<H", b, 2)
        self.ts_ns, = struct.unpack_from("<Q", b, 4)
        gm, gd = struct.unpack_from("<HI", b, 12)
        am, ad = struct.unpack_from("<HI", b, 27)
        i24 = lambda o: int.from_bytes(b[o:o + 3], "little", signed=True)
        self.gyro_dps = tuple(i24(18 + 3 * i) * gm / gd for i in range(3))
        self.accel_g = tuple(i24(33 + 3 * i) * am / ad for i in range(3))
        self.counter, = struct.unpack_from("<I", b, 54)


def _packet(cmd, data=b""):
    body = struct.pack("<HB", len(data) + 3, cmd) + data
    return bytes([0xAA]) + struct.pack("<I", zlib.crc32(body) & 0xFFFFFFFF) + body


# ---- MCU channel (interface 0, 0xFD framing) — device info queries -----------

def _mcu_packet(cmd, data=b""):
    body = struct.pack("<HIIH5s", len(data) + 17, 0x1337, 0, cmd, b"\0" * 5) + data
    return bytes([0xFD]) + struct.pack("<I", zlib.crc32(body) & 0xFFFFFFFF) + body


def mcu_query(cmd, data=b"", tries=50):
    """One read command on the MCU interface; returns (status, payload)."""
    path = next((d["path"] for d in hid.enumerate(VID, PID)
                 if d["interface_number"] == MCU_INTERFACE), None)
    if path is None:
        raise FileNotFoundError("MCU HID interface 0 not found")
    dev = hid.device()
    dev.open_path(path)
    try:
        dev.write(b"\x00" + _mcu_packet(cmd, data).ljust(64, b"\0"))
        for _ in range(tries):
            buf = dev.read(1024, timeout_ms=250)
            if not buf:
                return None
            b = bytes(buf)
            if b[:1] == b"\xfd":
                length, = struct.unpack_from("<H", b, 5)
                rcmd, = struct.unpack_from("<H", b, 15)
                if rcmd == cmd:
                    payload = b[22:22 + max(0, length - 17)]
                    return (payload[0] if payload else None,
                            payload[1:].decode("ascii", "replace").strip("\x00"))
        return None
    finally:
        dev.close()


def print_info():
    """Serial + firmware versions (command ids per the Air-family MCU
    protocol, wheaney/nrealAirLinuxDriver device_mcu.h)."""
    for label, cmd in (("serial       ", 0x15), ("MCU app FW   ", 0x26),
                       ("DP7911 FW    ", 0x16), ("DSP app FW   ", 0x21)):
        got = mcu_query(cmd)
        print(f"{label}: {got[1] if got else 'no reply'}")


class XrealImu:
    """The IMU HID channel. Use as a context manager; iterate samples()."""

    def __init__(self):
        path = next((d["path"] for d in hid.enumerate(VID, PID)
                     if d["interface_number"] == IMU_INTERFACE), None)
        if path is None:
            raise FileNotFoundError(
                "XREAL Air 2 Ultra HID interface 2 not found - glasses plugged in? "
                "(on Linux: udev rule or root needed)")
        self.dev = hid.device()
        self.dev.open_path(path)
        self._streaming = False

    def command(self, cmd, data=b"", tries=100):
        """Send one 0xAA command, return its reply payload (or None)."""
        self.dev.write(b"\x00" + _packet(cmd, data).ljust(64, b"\0"))
        for _ in range(tries):
            buf = self.dev.read(1024, timeout_ms=250)
            if not buf:
                return None
            b = bytes(buf)
            if b[:1] == b"\xaa":                      # command reply
                length, rcmd = struct.unpack_from("<HB", b, 5)
                if rcmd == cmd:
                    return b[8:8 + max(0, length - 3)]
        return None

    def stream(self, on):
        if self.command(0x19, b"\x01" if on else b"\x00") is None:
            raise RuntimeError("no ack for IMU stream command")
        self._streaming = on

    def config_json(self):
        """Fetch the factory calibration JSON stored on the glasses."""
        was = self._streaming
        if was:
            self.stream(False)
        raw = self.command(0x14)
        if raw is None or len(raw) < 4:
            raise RuntimeError("config length request failed")
        total, = struct.unpack("<I", raw[:4])
        cfg = bytearray()
        while len(cfg) < total:
            part = self.command(0x15)
            if part is None:
                raise RuntimeError(f"config read stalled at {len(cfg)}/{total}")
            cfg.extend(part)
        if was:
            self.stream(True)
        return bytes(cfg[:total])

    def samples(self):
        """Generator of ImuSample. Call stream(True) first."""
        while True:
            buf = self.dev.read(1024, timeout_ms=500)
            if not buf:
                yield None                            # stall marker
                continue
            b = bytes(buf)
            if b[:2] == SIG:
                yield ImuSample(b)

    def __enter__(self):
        return self

    def __exit__(self, *exc):
        # deliberately leave the stream running: it is free-running and
        # costless with no listener, and turning it off here would cut the
        # feed under any other process reading it in parallel
        self.dev.close()


class ImuRingWriter:
    """Publish samples to the named shared-memory ring (layout above)."""

    def __init__(self):
        from multiprocessing import shared_memory
        size = IMU_FB_HDR + IMU_FB_CAPACITY * IMU_FB_SLOT
        try:
            self.shm = shared_memory.SharedMemory(name=IMU_FB_NAME, create=True,
                                                  size=size)
        except FileExistsError:
            self.shm = shared_memory.SharedMemory(name=IMU_FB_NAME)
        struct.pack_into("<IIII", self.shm.buf, 0,
                         IMU_FB_MAGIC, 1, IMU_FB_CAPACITY, IMU_FB_SLOT)
        struct.pack_into("<Q", self.shm.buf, 16, 0)
        self.count = 0

    def publish(self, s):
        off = IMU_FB_HDR + (self.count % IMU_FB_CAPACITY) * IMU_FB_SLOT
        struct.pack_into("<QIHH3f3f4x", self.shm.buf, off,
                         s.ts_ns, s.counter, s.temp_raw, 0,
                         *s.gyro_dps, *s.accel_g)
        self.count += 1
        struct.pack_into("<Q", self.shm.buf, 16, self.count)  # publish after write

    def close(self):
        self.shm.close()
        try:
            self.shm.unlink()          # no-op on Windows
        except FileNotFoundError:
            pass


def main():
    ap = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    ap.add_argument("--csv", metavar="OUT.csv", help="log every sample")
    ap.add_argument("--fb", action="store_true",
                    help=f'publish to shared-memory ring "{IMU_FB_NAME}"')
    ap.add_argument("--config", metavar="OUT.json",
                    help="dump the factory calibration JSON and exit")
    ap.add_argument("--quat", action="store_true",
                    help="fuse a 6-axis quaternion host-side (Madgwick; the "
                         "device itself streams raw data only)")
    ap.add_argument("--info", action="store_true",
                    help="print serial + firmware versions (MCU channel) and exit")
    ap.add_argument("--seconds", type=float, help="stop after this long")
    args = ap.parse_args()

    if args.info:
        print_info()
        return

    with XrealImu() as imu:
        if args.config:
            cfg = imu.config_json()
            with open(args.config, "wb") as fh:
                fh.write(cfg)
            print(f"factory calibration ({len(cfg)} bytes) -> {args.config}")
            return

        csv = open(args.csv, "w") if args.csv else None
        if csv:
            csv.write("ts_ns,counter,temp_raw,gx_dps,gy_dps,gz_dps,ax_g,ay_g,az_g"
                      + (",qw,qx,qy,qz" if args.quat else "") + "\n")
        ring = ImuRingWriter() if args.fb else None
        if ring:
            print(f'publishing to shared memory "{IMU_FB_NAME}"')
        ahrs = None
        if args.quat:
            from xreal_ahrs import ImuOrientation, quat_to_euler_deg
            ahrs = ImuOrientation()
            print("estimating gyro bias, hold the glasses still for ~1 s...")

        imu.stream(True)
        print("IMU stream on (1 kHz). Ctrl+C to stop.")
        n, t0 = 0, time.time()
        end = time.time() + args.seconds if args.seconds else None
        try:
            for s in imu.samples():
                if end and time.time() > end:
                    break
                if s is None:
                    print("stream stalled...", flush=True)
                    continue
                n += 1
                q = ahrs.feed(s) if ahrs else None
                if csv:
                    csv.write(f"{s.ts_ns},{s.counter},{s.temp_raw},"
                              f"{s.gyro_dps[0]:.4f},{s.gyro_dps[1]:.4f},{s.gyro_dps[2]:.4f},"
                              f"{s.accel_g[0]:.5f},{s.accel_g[1]:.5f},{s.accel_g[2]:.5f}"
                              + (",%.6f,%.6f,%.6f,%.6f" % tuple(q) if q else
                                 ",,,," if ahrs else "") + "\n")
                if ring:
                    ring.publish(s)
                dt = time.time() - t0
                if dt >= 1:
                    if q:
                        yaw, pitch, roll = quat_to_euler_deg(q)
                        print(f"{n/dt:6.0f} Hz  q=({q[0]:+6.3f},{q[1]:+6.3f},{q[2]:+6.3f},"
                              f"{q[3]:+6.3f})  yaw={yaw:+7.1f} pitch={pitch:+6.1f} "
                              f"roll={roll:+7.1f} deg"
                              + ("" if ahrs.still_capture else "  [bias capture was noisy]"),
                              flush=True)
                    else:
                        a = s.accel_g
                        mag = (a[0] ** 2 + a[1] ** 2 + a[2] ** 2) ** 0.5
                        print(f"{n/dt:6.0f} Hz  gyro=({s.gyro_dps[0]:+8.3f},{s.gyro_dps[1]:+8.3f},"
                              f"{s.gyro_dps[2]:+8.3f}) deg/s  accel=({a[0]:+6.3f},{a[1]:+6.3f},"
                              f"{a[2]:+6.3f}) g  |a|={mag:5.3f}  t_raw={s.temp_raw}",
                              flush=True)
                    n, t0 = 0, time.time()
        except KeyboardInterrupt:
            print("\nstopped")
        finally:
            if csv:
                csv.close()
            if ring:
                ring.close()


if __name__ == "__main__":
    main()
